### DISCUSSION TITLE
warning on foreach conversions

### DISCUSSION THREAD

**Discussion by ZacharyPatten** [dotnet/roslyn#57308] [Ideas] [answered]
> _**NOTE! This conversation has been moved to roslyn-analyzers. Please continue discussions there: https://github.com/dotnet/roslyn-analyzers/issues/5672**_

There should be a compiler warning when a `foreach` loop compiles and adds an `explicit` cast operation. Here is an example that would apply:
```cs
// Example #1: foreach adds explicits casts between enum types
using System;

foreach (A a in Enum.GetValues<B>()) // <- warning
{
	Console.WriteLine(a);
}

enum A { A = 'a', B = 'b', C = 'c', }
enum B { D = 'd', E = 'e', F = 'f', }

// Output:
// 100
// 101
// 102
```

Another example:
```cs
// Example #2: foreach adds explicits casts long -> int
long[] array = { long.MinValue, 1, 2, 3, long.MaxValue };
foreach (int i in array) // <- warning
{
	Console.WriteLine(i);
}
// Output:
// 0
// 1
// 2
// 3
// -1
```

From my experience, this behavior often results in bugs in code. I honeslty prefer this code not compile, but if there was at least a warning then I can just increase the severity of that warning to be an error and thus break builds.

## Acknowledging This May Be An Annoying Warning

If this warning was added there currently isn't any clean syntax to explicitly allow explicit casts in `foreach` loops. So develpoers who cannot simply ignore or suppress it may find this warning annoying. Here are two examples that modify `Example #2` so that it would not result in the proposed warning:

```cs
// Example #3: modify Example #2 to bypass warning via casting in loop
long[] array = { long.MinValue, 1, 2, 3, long.MaxValue };
foreach (long i in array)
{
	int i2 = (int)i;
	Console.WriteLine(i2);
}
```

```cs
// Example #4: modify Example #2 to bypass warning via Linq Select
long[] array = { long.MinValue, 1, 2, 3, long.MaxValue };
foreach (int i in array.Select(l => (int)l))
{
	Console.WriteLine(i);
}
```

So while there currently would be ways to bypass the warning via syntax (rather than suppressing/ignoring it), those ways are not really ideal.

## More Good Than Harm

In my opinion, adding this compiler warning would do more good than harm. I would assume intentionally allowing explicit casts from `foreach` loops is relatively uncommon and accidentally using the wrong type in a foreach loop is relatively common (especially for new developers). Without a compiler warning these bugs can sometimes be hard to find. With a compiler warning (that we can escalate to an error if we choose), we can prevent these bugs with ease.

## What About Implicit Casts?

It may be nice if there was a warning for implicit casts too (not just explicit casts as mentioned in this post). If that was considered, it would be nice if it was a seperate warning so we can configure implicit and explicit casts seperately. Also, for implicit casts, maybe that would be better as a compiler message than a warning since it would seem less likely to be a bug in the code.

**Comment by jmarolf** [comment, marked as answer]
Why does this need to be in the compiler? Why can it not be an analyzer that folks who know this is an error in their codebase can choose to install. There are millions of lines of C# code today that we working just fine while casting in `foreach`. I also do not think we want to warn for behavior that is defined [in the spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#the-foreach-statement).

The C# spec says the following

> A foreach statement of the form
```csharp
foreach (V v in x) embedded_statement
```
> is then expanded to:
```csharp
{
    E e = ((C)(x)).GetEnumerator();
    try {
        while (e.MoveNext()) {
            V v = (V)(T)e.Current;
            embedded_statement
        }
    }
    finally {
        ... // Dispose e
    }
}
```

As long as the C# spec defines foreach as explicitly casting by design I do not think the compiler should warn (though someone is free to implement an analyzer on their own).

**Reply by ZacharyPatten** [reply]
I opened the discussion here after a brief mention on the C# discord. There was some doubt that the necessary information for an analyzer to detect this topic is exposed.

If it can be implemented in https://github.com/dotnet/roslyn-analyzers then that sounds good to me. :)

**Reply by CyrusNajmabadi** [reply]
Hrmm.. i can't seem to transfer this.  Weird...

**Reply by ZacharyPatten** [reply]
If it should be transferred and you can't feel free to close it and I can open a new discussion in https://github.com/dotnet/roslyn-analyzers.

@333fred any comments before this is moved though?

**Reply by 333fred** [reply]
I don't think this can currently be implemented in an analyzer. Mainly for technical reasons, as I don't believe we expose the necessary information.

**Reply by CyrusNajmabadi** [reply]
Can you give examples @333fred ?  For example, we expose all of:

```
    public struct ForEachStatementInfo : IEquatable<ForEachStatementInfo>
    {
        public bool IsAsynchronous { get; }
        public IMethodSymbol? GetEnumeratorMethod { get; }
        public IMethodSymbol? MoveNextMethod { get; }
        public IPropertySymbol? CurrentProperty { get; }
        public IMethodSymbol? DisposeMethod { get; }
        public ITypeSymbol? ElementType { get; }
        public Conversion ElementConversion { get; }
        public Conversion CurrentConversion { get; }
```

Are those not the relevant conversions?  or is there something else missing?  Thanks!

**Reply by 333fred** [reply]
My search-fu must not be up to snuff, I totally missed that struct. That is the info I thought we didn't expose.

**Reply by RikkiGibson** [reply]
> Hrmm.. i can't seem to transfer this. Weird...

It looks like roslyn-analyzers doesn't have discussions enabled. Maybe it would be best for the OP to open an issue in roslyn-analyzers and link to it here.

**Reply by ZacharyPatten** [reply]
> Maybe it would be best for the OP to open an issue in roslyn-analyzers and link to it here.

https://github.com/dotnet/roslyn-analyzers/issues/5672

**Reply by mavasani** [reply]
> As long as the C# spec defines foreach as explicitly casting by design I do not think the compiler should warn (though someone is free to implement an analyzer on their own).

I agree with @jmarolf here and also stated this concern in https://github.com/dotnet/roslyn/issues/33286. This should absolutely not be a compiler warning, and I would be very wary of it be an SDK analyzer suggestion/warning that flags use of a particular C# language feature in the IDE and recommending developers to not use it. At the very best, it can be a hidden code style diagnostic with a code fix to allow users to switch to the explicit cast form. I think it might be best if this were to be implemented as a 3rd party analyzer rather than something shipping in the box given that it pushes against using a language feature.

**Comment by RikkiGibson** [comment]
The main concern I have about foreach conversions is the difficulty of suppressing the warning.

```cs
class Base { }
class Derived : Base { }

void M(List<Base> items)
{
    foreach (Derived derived in items)
    {
    }
}
```

Basically, what happens if I really do want the implicit cast here? Should I have to declare a local and assign a cast of the foreach variable to it?

That said, we have had [functional bugs](https://github.com/dotnet/roslyn/pull/56233/files#diff-fd41468c64dd070639cc4e760f991439fbfe752874a14791d981ca127945d7fa) in Roslyn due to the implicit cast behavior. I honestly think it might be good to include it as a warning out of the box (i.e. either in compiler or SDK analyzers) in .NET 7.

**Reply by jasonmalinowski** [reply]
My gut is this behavior comes back from the C# 1.0 days where you didn't have generics and IEnumerable only gave you an object...if you didn't have this behavior it would have made that pretty painful.

If you really want the type change, there's always `foreach (Derived derived in items.Cast<Derived>())` which makes the cast explicit.

**Reply by CyrusNajmabadi** [reply]
unfortunately, that adds perf overhead.  Iterating a List<Base> is cheaper than the casted enumerable produced by .Cast<>.

**Reply by CyrusNajmabadi** [reply]
> Basically, what happens if I really do want the implicit cast here? Should I have to declare a local and assign a cast of the foreach variable to it?

TBH, yes, i think that's reasonable.  It's unclear to me why this behavior is ever a good thing (now that we have generics and whatnot).  In all other circumstances we don't implicitly downcast (afaik).  Instead, we make you do it explicitly.  And that's a good thing as downcasts should be explicit in teh code in some fashion IMO.

**Reply by ZacharyPatten** [reply]
> we have had functional bugs in Roslyn due to the implicit cast behavior

I have also encountered multiple bugs caused by this issue that made it into production code.

Also, I have helped a lot of new developers learn C#, and I have watched many of them encounter this issue. They will change the element type of an array or list and just forget to change the iteration variable of the `foreach` loop. The code will still compile but things seem odd, and (being in the early stages of learning C#) they could be scratching their head for a long time before realizing the issue. So besides preventing bugs in production code, a compiler warning will also help guide beginners learning C#.

**Reply by mavasani** [reply]
Such an analyzer was also discussed and turned down in https://github.com/dotnet/runtime/issues/48529

**Reply by jasonmalinowski** [reply]
Looking at the conversation it looks like there's absolutely people who _do_ want it, but yes, it's definitely likely to be noisy. @mavasani do we have other precedent for writing an analyzer that's off by default?

**Reply by jmarolf** [reply]
This feels like the IDisposable analyzer that we ship [CA1001](https://github.com/dotnet/roslyn-analyzers/blob/6ae50529e7b88493b2591fb533ceebadc8425e1d/src/NetAnalyzers/Core/AnalyzerReleases.Shipped.md?plain=1#L8) which also can't be implemented such that false positives are impossible. This seems like a reasonable path forward: implement it in roslyn-analyzers. ship it. but have it be "hidden" by default.
